### PR TITLE
feat(route): 山东大学（威海）新闻网

### PR DESCRIPTION
**Pull request by kxxt**
## 完整路由地址 / Example for the proposed route(s)

```routes
/sdu/wh/news/
/sdu/wh/news/xyyw
/sdu/wh/news/xsdt
/sdu/wh/news/zhxw
/sdu/wh/news/sdsd
/sdu/wh/news/jjxy
/sdu/wh/news/xyjx
/sdu/wh/news/mjzc
/sdu/wh/news/rdzt
/sdu/wh/news/mtsj
/sdu/wh/news/gjsy
/sdu/wh/news/llxx
```

## 新 RSS 检查列表 / New RSS Script Checklist
  
- [x] 新的路由 New Route
  - [x] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [x] 文档说明 Documentation
  - [x] 中文文档 CN
  - [ ] 英文文档 EN
- [x] 全文获取 fulltext
  - [x] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [x] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [x] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note

我并不太清楚 `radar.js` 配置的是否正确，因为我在自己调试时没有成功。

我在将以下配置复制到 RSS Radar 配置界面后，打开 https://xinwen.wh.sdu.edu.cn/ 之后， RSS Radar 并没有检测到 RSS.
我没有找到原因，还请 reviewer 指点
<details>
<summary>生成的配置</summary>

```js
{
  'sdu.edu.cn': {
    _name: '山东大学',
    'xinwen.wh': [
      {
        title: '校园要闻',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/xyyw'
      },
      {
        title: '学生动态',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/xsdt'
      },
      {
        title: '综合新闻',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/zhxw'
      },
      {
        title: '山大视点',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/sdsd'
      },
      {
        title: '菁菁校园',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/jjxy'
      },
      {
        title: '校园简讯',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/xyjx'
      },
      {
        title: '玛珈之窗',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/mjzc'
      },
      {
        title: '热点专题',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/rdzt'
      },
      {
        title: '媒体视角',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/mtsj'
      },
      {
        title: '高教视野',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/gjsy'
      },
      {
        title: '理论学习',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/llxx'
      }
    ]
  }
}
```
</detail>

### CODE REVIEW

**Comment by TonyRL** [comment]
```JSON
      {
        title: '校园要闻',
        docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
        source: [ '/' ],
        target: '/sduwh/news/xyyw'
      }
```
Having source '/' will match `example.com/` only, you need to use wildcard match `*` in this case, e.g., `/*.html`

File: lib/v2/sduwh/extractor/news.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+
+module.exports = async (link, ctx) =>
+    await ctx.cache.tryGet(link, async () => {
+        try {
+            const result = await got.get(link);
```
Comment (TonyRL) [inline lib/v2/sduwh/extractor/news.js:7]:
```suggestion
            const result = await got(link);
```

File: lib/v2/sduwh/extractor/view.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+
+module.exports = async (link, ctx) =>
+    await ctx.cache.tryGet(link, async () => {
+        try {
+            const result = await got.get(link);
```
Comment (TonyRL) [inline lib/v2/sduwh/extractor/view.js:7]:
```suggestion
            const result = await got(link);
```

File: lib/v2/sduwh/news.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const data = require('./data').news;
+const extractor = require('./extractor');
+
+module.exports = async (ctx) => {
+    const column = ctx.params.column ?? 'xyyw';
+    const baseUrl = data.url;
+    const response = await got.get(baseUrl + data.columns[column].url);
```
Comment (TonyRL) [inline lib/v2/sduwh/news.js:9]:
```suggestion
    const response = await got(baseUrl + data.columns[column].url);
```

File: lib/v2/sduwh/news.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    const out = await Promise.all(
+        items.map(async (index, item) => {
+            item = $(item);
+            const anchor = item.find('a');
+            const title = anchor.attr('title');
+            const href = anchor.attr('href');
+            const link = href.startsWith('http') ? href : baseUrl + '/' + href;
+            const { description, author } = await ctx.cache.tryGet(link, async () => await extractor(link, ctx));
+            const span = item.find('span');
+            const pubDate = new Date(span.text());
```
Comment (TonyRL) [inline lib/v2/sduwh/news.js:21]:
```suggestion
            const pubDate = parseDate(span.text(), 'YYYY/MM/DD');
```
https://docs.rsshub.app/joinus/pub-date.html#shi-yong-gong-ju-lei-parse-date

---
Btw articles have the exact pubDate and time inside `.news_tit`, consider using it.

**Review by TonyRL — COMMENTED** [review]
Consider using the route `/sdu/wh/news/*` instead

File: lib/v2/sduwh/extractor/news.js (hunk 1 above)
Comment (kxxt) [reply lib/v2/sduwh/extractor/news.js]: Fixed now. Thanks for pointing out this.

**Comment by kxxt** [comment]
> Consider using the route `/sdu/wh/news/*` instead

I noticed that the `/sdu` route hasn't been migrated to v2 yet.
If I change my route(v2) to `/sdu/wh/news`, will the old route conflict with my new route? (I'm not familiar with koa routers)

**Comment by TonyRL** [comment]
> > Consider using the route `/sdu/wh/news/*` instead
> 
> I noticed that the `/sdu` route hasn't been migrated to v2 yet. If I change my route(v2) to `/sdu/wh/news`, will the old route conflict with my new route? (I'm not familiar with koa routers)

No. It won't conflict with your routes.

**Comment by kxxt** [comment]
> > > Consider using the route `/sdu/wh/news/*` instead
> > 
> > 
> > I noticed that the `/sdu` route hasn't been migrated to v2 yet. If I change my route(v2) to `/sdu/wh/news`, will the old route conflict with my new route? (I'm not familiar with koa routers)
> 
> No. It won't conflict with your routes.

Thank you. I'll change the route to `/sdu/wh/news`.

**Comment by kxxt** [comment]
> ```json
>       {
>         title: '校园要闻',
>         docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
>         source: [ '/' ],
>         target: '/sduwh/news/xyyw'
>       }
> ```
> 
> Having source '/' will match `/` only, you need to use wildcard match `*` in this case, e.g., `/*.html`

Just found that it works in Chromium but not in Firefox. 
It seems that there is a bug in the Firefox version of RSS Radar. 
https://github.com/DIYgod/RSSHub-Radar/issues/687

File: lib/v2/sdu/radar.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+const { news } = require('./data');
+
+module.exports = {
+    'sdu.edu.cn': {
+        _name: '山东大学',
```
Comment (kxxt) [inline lib/v2/sdu/radar.js:5]: By the way, any elegant way to set the subdomain's site name? (wh.sdu.edu.cn)
image
I could only come up a less elegant approach. (prepending `"（威海）新闻网 "` to each item's title)

Comment (TonyRL) [reply lib/v2/sdu/radar.js:5]: Possibly no. I don't think `_name` and `title` allow using function.

Comment (TonyRL) [reply lib/v2/sdu/radar.js:5]: Possibly no. I don't think `_name` and `title` allow using function.

Comment (TonyRL) [reply lib/v2/sdu/radar.js:5]: Btw if source is `/*.htm`
You will see this instead
<details>
<summary>Screenshot</summary>

image
</details>

Comment (kxxt) [reply lib/v2/sdu/radar.js:5]: Thanks. Then I'm going to use my less elegant approach.

**Comment by kxxt** [comment]
I think everything works fine now. Maybe we can get it merged.
Thank you very much for your review. This is actually the first time that I get a true review ever since I joined GitHub.

File: lib/v2/sdu/data.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+module.exports = {
+    wh: {
+        news: {
+            name: '山东大学（威海）新闻网',
+            route: '/news',
+            source: ['/'],
+            titlePrefix: '(威海)新闻网|',
+            docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai-xin-wen-wang',
```
Comment (TonyRL) [inline lib/v2/sdu/data.js:8]:
```suggestion
            docs: 'https://docs.rsshub.app/university.html#shan-dong-da-xue-wei-hai',
```
This should point to H2 heading

File: lib/v2/sdu/wh/news.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                description,
+                pubDate,
+                author,
+            };
+        })
+    );
+
+    ctx.state.data = {
+        title: `${data.name} ${data.columns[column].name}`,
+        link: data.url,
```
Comment (TonyRL) [inline lib/v2/sdu/wh/news.js:35]:
```suggestion
        link: baseUrl + data.columns[column].url,
```

File: lib/v2/sdu/extractor/sdrj.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+module.exports = async (link, ctx) =>
+    await ctx.cache.tryGet(link, async () => {
+        try {
+            const result = await got(link);
+            const $ = cheerio.load(result.data);
+            const content = $('#vsb_content').html();
+            const author = $("form[name='_newscontent_fromname'] > h1").text();
+            const exactDateLine = $("form[name='_newscontent_fromname'] > p.info").text().trim();
+            const exactDateText = exactDateLine.match(/^发布时间：(?<date>\d+\/\d+\/\d+\s\d{2}:\d{2}:\d{2})/).groups.date;
+            const exactDate = parseDate(exactDateText, 'YYYY/MM/DD HH:mm:ss');
```
Comment (TonyRL) [inline lib/v2/sdu/extractor/sdrj.js:14]:
```suggestion
            const exactDate = timezone(parseDate(exactDateText, 'YYYY/MM/DD HH:mm:ss'), +8);
```
You also need to adjust timezone.
`const timezone = require('@/utils/timezone');`

File: lib/v2/sdu/extractor/view.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+module.exports = async (link, ctx) =>
+    await ctx.cache.tryGet(link, async () => {
+        try {
+            const result = await got(link);
+            const $ = cheerio.load(result.data);
+            const content = $('#vsb_content').html();
+            const author = $("form[name='_newscontent_fromname'] > div > p:last-of-type").text();
+            const exactDateLine = $('.news_tit > p:last-child').text();
+            const exactDateText = exactDateLine.match(/^发布日期：(?<date>\d+年\d+月\d+日\s\d{2}:\d{2})/).groups.date;
+            const exactDate = parseDate(exactDateText, 'YYYY年MM月DD日 HH:mm');
```
Comment (TonyRL) [inline lib/v2/sdu/extractor/view.js:14]:
```suggestion
            const exactDate = timezone(parseDate(exactDateText, 'YYYY年MM月DD日 HH:mm'), +8);
```

File: lib/v2/sdu/extractor/wh/news.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+module.exports = async (link, ctx) =>
+    await ctx.cache.tryGet(link, async () => {
+        try {
+            const result = await got(link);
+            const $ = cheerio.load(result.data);
+            const content = $('#main-content').html();
+            const author = $('#source').text();
+            const exactDateLine = $('.news_tit > p:last-child').text();
+            const exactDateText = exactDateLine.match(/^发布日期：(?<date>\d+年\d+月\d+日\s\d{2}:\d{2})/).groups.date;
+            const exactDate = parseDate(exactDateText, 'YYYY年MM月DD日 HH:mm');
```
Comment (TonyRL) [inline lib/v2/sdu/extractor/wh/news.js:14]:
```suggestion
            const exactDate = timezone(parseDate(exactDateText, 'YYYY年MM月DD日 HH:mm'), +8);
```

**Pull request merged**